File: clients/php/Model/WeaponProperties.php
<?php

namespace Model {
    require_once 'Stream.php';

    /**
     * Weapon properties
     */
    class WeaponProperties
    {
        /**
         * Name
         */
        public string $name;
        /**
         * Shooting speed (number of shots per second)
         */
        public float $roundsPerSecond;
        /**
         * Accuracy (spread angle) of a shot (in degrees)
         */
        public float $spread;
        /**
         * Aiming time
         */
        public float $aimTime;
        /**
         * Field of view in full aim (in degrees)
         */
        public float $aimFieldOfView;
        /**
         * Rotation speed in full aim (degrees per second)
         */
        public float $aimRotationSpeed;
        /**
         * Movement speed modifier in full aim
         */
        public float $aimMovementSpeedModifier;
        /**
         * Speed of projectiles
         */
        public float $projectileSpeed;
        /**
         * Damage of a projectile
         */
        public float $projectileDamage;
        /**
         * Projectiles' life time
         */
        public float $projectileLifeTime;
        /**
         * Index of the sound when shooting (starting with 0), or None
         */
        public ?int $shotSoundTypeIndex;
        /**
         * Index of the sound when hitting something (starting with 0), or None
         */
        public ?int $projectileHitSoundTypeIndex;
        /**
         * Max amount of ammo unit can hold in their inventory
         */
        public int $maxInventoryAmmo;
    
        function __construct(string $name, float $roundsPerSecond, float $spread, float $aimTime, float $aimFieldOfView, float $aimRotationSpeed, float $aimMovementSpeedModifier, float $projectileSpeed, float $projectileDamage, float $projectileLifeTime, ?int $shotSoundTypeIndex, ?int $projectileHitSoundTypeIndex, int $maxInventoryAmmo)
        {
            $this->name = $name;
            $this->roundsPerSecond = $roundsPerSecond;
            $this->spread = $spread;
            $this->aimTime = $aimTime;
            $this->aimFieldOfView = $aimFieldOfView;
            $this->aimRotationSpeed = $aimRotationSpeed;
            $this->aimMovementSpeedModifier = $aimMovementSpeedModifier;
            $this->projectileSpeed = $projectileSpeed;
            $this->projectileDamage = $projectileDamage;
            $this->projectileLifeTime = $projectileLifeTime;
            $this->shotSoundTypeIndex = $shotSoundTypeIndex;
            $this->projectileHitSoundTypeIndex = $projectileHitSoundTypeIndex;
            $this->maxInventoryAmmo = $maxInventoryAmmo;
        }
    
        /**
         * Read WeaponProperties from input stream
         */
        public static function readFrom(\InputStream $stream): WeaponProperties
        {
            $name = $stream->readString();
            $roundsPerSecond = $stream->readDouble();
            $spread = $stream->readDouble();
            $aimTime = $stream->readDouble();
            $aimFieldOfView = $stream->readDouble();
            $aimRotationSpeed = $stream->readDouble();
            $aimMovementSpeedModifier = $stream->readDouble();
            $projectileSpeed = $stream->readDouble();
            $projectileDamage = $stream->readDouble();
            $projectileLifeTime = $stream->readDouble();
            if ($stream->readBool()) {
                $shotSoundTypeIndex = $stream->readInt32();
            } else {
                $shotSoundTypeIndex = NULL;
            }
            if ($stream->readBool()) {
                $projectileHitSoundTypeIndex = $stream->readInt32();
            } else {
                $projectileHitSoundTypeIndex = NULL;
            }
            $maxInventoryAmmo = $stream->readInt32();
            return new WeaponProperties($name, $roundsPerSecond, $spread, $aimTime, $aimFieldOfView, $aimRotationSpeed, $aimMovementSpeedModifier, $projectileSpeed, $projectileDamage, $projectileLifeTime, $shotSoundTypeIndex, $projectileHitSoundTypeIndex, $maxInventoryAmmo);
        }
        
        /**
         * Write WeaponProperties to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeString($this->name);
            $stream->writeDouble($this->roundsPerSecond);
            $stream->writeDouble($this->spread);
            $stream->writeDouble($this->aimTime);
            $stream->writeDouble($this->aimFieldOfView);
            $stream->writeDouble($this->aimRotationSpeed);
            $stream->writeDouble($this->aimMovementSpeedModifier);
            $stream->writeDouble($this->projectileSpeed);
            $stream->writeDouble($this->projectileDamage);
            $stream->writeDouble($this->projectileLifeTime);
            if (is_null($this->shotSoundTypeIndex)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $stream->writeInt32($this->shotSoundTypeIndex);
            }
            if (is_null($this->projectileHitSoundTypeIndex)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $stream->writeInt32($this->projectileHitSoundTypeIndex);
            }
            $stream->writeInt32($this->maxInventoryAmmo);
        }
    }
}

File: clients/go/model/weapon_properties.go
package model

import (
	"fmt"
	"io"

	. "aicup22/stream"
)

// Weapon properties
type WeaponProperties struct {
	// Name
	Name string
	// Shooting speed (number of shots per second)
	RoundsPerSecond float64
	// Accuracy (spread angle) of a shot (in degrees)
	Spread float64
	// Aiming time
	AimTime float64
	// Field of view in full aim (in degrees)
	AimFieldOfView float64
	// Rotation speed in full aim (degrees per second)
	AimRotationSpeed float64
	// Movement speed modifier in full aim
	AimMovementSpeedModifier float64
	// Speed of projectiles
	ProjectileSpeed float64
	// Damage of a projectile
	ProjectileDamage float64
	// Projectiles' life time
	ProjectileLifeTime float64
	// Index of the sound when shooting (starting with 0), or None
	ShotSoundTypeIndex *int32
	// Index of the sound when hitting something (starting with 0), or None
	ProjectileHitSoundTypeIndex *int32
	// Max amount of ammo unit can hold in their inventory
	MaxInventoryAmmo int32
}

func NewWeaponProperties(name string, roundsPerSecond float64, spread float64, aimTime float64, aimFieldOfView float64, aimRotationSpeed float64, aimMovementSpeedModifier float64, projectileSpeed float64, projectileDamage float64, projectileLifeTime float64, shotSoundTypeIndex *int32, projectileHitSoundTypeIndex *int32, maxInventoryAmmo int32) WeaponProperties {
	return WeaponProperties{
		Name:                        name,
		RoundsPerSecond:             roundsPerSecond,
		Spread:                      spread,
		AimTime:                     aimTime,
		AimFieldOfView:              aimFieldOfView,
		AimRotationSpeed:            aimRotationSpeed,
		AimMovementSpeedModifier:    aimMovementSpeedModifier,
		ProjectileSpeed:             projectileSpeed,
		ProjectileDamage:            projectileDamage,
		ProjectileLifeTime:          projectileLifeTime,
		ShotSoundTypeIndex:          shotSoundTypeIndex,
		ProjectileHitSoundTypeIndex: projectileHitSoundTypeIndex,
		MaxInventoryAmmo:            maxInventoryAmmo,
	}
}

// Read WeaponProperties from reader
func ReadWeaponProperties(reader io.Reader) WeaponProperties {
	var name string
	name = ReadString(reader)
	var roundsPerSecond float64
	roundsPerSecond = ReadFloat64(reader)
	var spread float64
	spread = ReadFloat64(reader)
	var aimTime float64
	aimTime = ReadFloat64(reader)
	var aimFieldOfView float64
	aimFieldOfView = ReadFloat64(reader)
	var aimRotationSpeed float64
	aimRotationSpeed = ReadFloat64(reader)
	var aimMovementSpeedModifier float64
	aimMovementSpeedModifier = ReadFloat64(reader)
	var projectileSpeed float64
	projectileSpeed = ReadFloat64(reader)
	var projectileDamage float64
	projectileDamage = ReadFloat64(reader)
	var projectileLifeTime float64
	projectileLifeTime = ReadFloat64(reader)
	var shotSoundTypeIndex *int32
	if ReadBool(reader) {
		var shotSoundTypeIndexValue int32
		shotSoundTypeIndexValue = ReadInt32(reader)
		shotSoundTypeIndex = &shotSoundTypeIndexValue
	} else {
		shotSoundTypeIndex = nil
	}
	var projectileHitSoundTypeIndex *int32
	if ReadBool(reader) {
		var projectileHitSoundTypeIndexValue int32
		projectileHitSoundTypeIndexValue = ReadInt32(reader)
		projectileHitSoundTypeIndex = &projectileHitSoundTypeIndexValue
	} else {
		projectileHitSoundTypeIndex = nil
	}
	var maxInventoryAmmo int32
	maxInventoryAmmo = ReadInt32(reader)
	return WeaponProperties{
		Name:                        name,
		RoundsPerSecond:             roundsPerSecond,
		Spread:                      spread,
		AimTime:                     aimTime,
		AimFieldOfView:              aimFieldOfView,
		AimRotationSpeed:            aimRotationSpeed,
		AimMovementSpeedModifier:    aimMovementSpeedModifier,
		ProjectileSpeed:             projectileSpeed,
		ProjectileDamage:            projectileDamage,
		ProjectileLifeTime:          projectileLifeTime,
		ShotSoundTypeIndex:          shotSoundTypeIndex,
		ProjectileHitSoundTypeIndex: projectileHitSoundTypeIndex,
		MaxInventoryAmmo:            maxInventoryAmmo,
	}
}

// Write WeaponProperties to writer
func (weaponProperties WeaponProperties) Write(writer io.Writer) {
	name := weaponProperties.Name
	WriteString(writer, name)
	roundsPerSecond := weaponProperties.RoundsPerSecond
	WriteFloat64(writer, roundsPerSecond)
	spread := weaponProperties.Spread
	WriteFloat64(writer, spread)
	aimTime := weaponProperties.AimTime
	WriteFloat64(writer, aimTime)
	aimFieldOfView := weaponProperties.AimFieldOfView
	WriteFloat64(writer, aimFieldOfView)
	aimRotationSpeed := weaponProperties.AimRotationSpeed
	WriteFloat64(writer, aimRotationSpeed)
	aimMovementSpeedModifier := weaponProperties.AimMovementSpeedModifier
	WriteFloat64(writer, aimMovementSpeedModifier)
	projectileSpeed := weaponProperties.ProjectileSpeed
	WriteFloat64(writer, projectileSpeed)
	projectileDamage := weaponProperties.ProjectileDamage
	WriteFloat64(writer, projectileDamage)
	projectileLifeTime := weaponProperties.ProjectileLifeTime
	WriteFloat64(writer, projectileLifeTime)
	shotSoundTypeIndex := weaponProperties.ShotSoundTypeIndex
	if shotSoundTypeIndex == nil {
		WriteBool(writer, false)
	} else {
		WriteBool(writer, true)
		shotSoundTypeIndexValue := *shotSoundTypeIndex
		WriteInt32(writer, shotSoundTypeIndexValue)
	}
	projectileHitSoundTypeIndex := weaponProperties.ProjectileHitSoundTypeIndex
	if projectileHitSoundTypeIndex == nil {
		WriteBool(writer, false)
	} else {
		WriteBool(writer, true)
		projectileHitSoundTypeIndexValue := *projectileHitSoundTypeIndex
		WriteInt32(writer, projectileHitSoundTypeIndexValue)
	}
	maxInventoryAmmo := weaponProperties.MaxInventoryAmmo
	WriteInt32(writer, maxInventoryAmmo)
}

// Get string representation of WeaponProperties
func (weaponProperties WeaponProperties) String() string {
	stringResult := "{ "
	stringResult += "Name: "
	name := weaponProperties.Name
	stringResult += "\"" + name + "\""
	stringResult += ", "
	stringResult += "RoundsPerSecond: "
	roundsPerSecond := weaponProperties.RoundsPerSecond
	stringResult += fmt.Sprint(roundsPerSecond)
	stringResult += ", "
	stringResult += "Spread: "
	spread := weaponProperties.Spread
	stringResult += fmt.Sprint(spread)
	stringResult += ", "
	stringResult += "AimTime: "
	aimTime := weaponProperties.AimTime
	stringResult += fmt.Sprint(aimTime)
	stringResult += ", "
	stringResult += "AimFieldOfView: "
	aimFieldOfView := weaponProperties.AimFieldOfView
	stringResult += fmt.Sprint(aimFieldOfView)
	stringResult += ", "
	stringResult += "AimRotationSpeed: "
	aimRotationSpeed := weaponProperties.AimRotationSpeed
	stringResult += fmt.Sprint(aimRotationSpeed)
	stringResult += ", "
	stringResult += "AimMovementSpeedModifier: "
	aimMovementSpeedModifier := weaponProperties.AimMovementSpeedModifier
	stringResult += fmt.Sprint(aimMovementSpeedModifier)
	stringResult += ", "
	stringResult += "ProjectileSpeed: "
	projectileSpeed := weaponProperties.ProjectileSpeed
	stringResult += fmt.Sprint(projectileSpeed)
	stringResult += ", "
	stringResult += "ProjectileDamage: "
	projectileDamage := weaponProperties.ProjectileDamage
	stringResult += fmt.Sprint(projectileDamage)
	stringResult += ", "
	stringResult += "ProjectileLifeTime: "
	projectileLifeTime := weaponProperties.ProjectileLifeTime
	stringResult += fmt.Sprint(projectileLifeTime)
	stringResult += ", "
	stringResult += "ShotSoundTypeIndex: "
	shotSoundTypeIndex := weaponProperties.ShotSoundTypeIndex
	if shotSoundTypeIndex == nil {
		stringResult += "nil"
	} else {
		shotSoundTypeIndexValue := *shotSoundTypeIndex
		stringResult += fmt.Sprint(shotSoundTypeIndexValue)
	}
	stringResult += ", "
	stringResult += "ProjectileHitSoundTypeIndex: "
	projectileHitSoundTypeIndex := weaponProperties.ProjectileHitSoundTypeIndex
	if projectileHitSoundTypeIndex == nil {
		stringResult += "nil"
	} else {
		projectileHitSoundTypeIndexValue := *projectileHitSoundTypeIndex
		stringResult += fmt.Sprint(projectileHitSoundTypeIndexValue)
	}
	stringResult += ", "
	stringResult += "MaxInventoryAmmo: "
	maxInventoryAmmo := weaponProperties.MaxInventoryAmmo
	stringResult += fmt.Sprint(maxInventoryAmmo)
	stringResult += " }"
	return stringResult
}
